### ISSUE TITLE
hostname changing to localhost (SLES 12)

### ISSUE THREAD

**Issue by jakehoms** [Azure/WALinuxAgent#1399]
Please see [this link](https://github.com/Azure/WALinuxAgent/issues/742) as reference to the exact issue I am having, but it was closed last year due to no follow-up by the original poster. I am trying to create a SLES 12 for SAP SP3 instance in Azure.

Few extra details: We do this process for our RHEL/CentOS/OEL images just fine, but our Suse images do not work and re-set the hostname to localhost, even after setting it correctly earlier in boot. I am creating the image using the GUI of Azure and cloud-init is not being used (as it's not supported in Suse). This is solely using WALinuxAgent. Please let me know if you need more info. Thanks. 

This is the link to my waagent.log in verbose mode (the first portion which covers the hostname changes). 

https://pastebin.com/pDaRFuGv
Lines 66-68: Set hostname as chosen from the Azure GUI (desired state)
Lines 343-344: Recognize and set hostname to localhost  (not the desired state)

_Originally posted by @jakehoms in https://github.com/Azure/WALinuxAgent/issue_comments#issuecomment-439092149_

**Comment by Darryl-G** [comment]
Are you changing any entries in the following or using a custom script (via WAAgent) to change entries in:
- /etc/hosts
- /etc/resolv.conf
- /etc/sysconfig/network/config
?

I have SLES 12 for SAP SP3 and our hostnames are coming through without issue.

**Comment by jakehoms** [comment]
Hi Darryl-G. We are building from ISO but are not making any changes to /etc/hosts or /etc/resolv.conf, only allowing DHCP to handle the latter. For /etc/sysconfig/network/config, we have tested numerous changes (from no changes and leaving it as default, to adjusting the DNS servers and Modules order) but all failed to keep the hostname the same and ultimately put it back to localhost. Are you building from ISO or Marketplace image? I can confirm that the Marketplace image does work as expected, keeping the hostname, but the ISO build does not and is required by our team. We have the same pipeline for the other RHEL OS types and they are functioning as expected with the hostname change.

**Comment by Darryl-G** [comment]
I'm building from marketplace.
There are subtle differences in the contents of packages in the ISO vs marketplace.
Do you have a private IP address on eth0 ?  If yes, check the reverse lookup for this IP address from the server itself ?
Also might be worth checking the SLES firewall is not blocking the WAAgent and especially the metadata service on the 192.x address.  Are you able to query the WAAgent metadata?
The Azure metadata is used by cloud-netconfig in SLES 12 SP3 to set IP, DNS, routing.
p.s. I'm only on here because I've got an issue I'm troubleshooting with cloud-netconfig.

**Comment by jakehoms** [comment]
Thanks again Darryl-G for the info! Some great suggestions and I actually walked through most of them earlier. Thankfully, I found the resolution in the meantime! We have a shared environment and one of the other people working on this build did setup a hostname config in the AutoYast xml config. Example here:

     </dhcp_options>
     <dns>
       <dhcp_hostname config:type="boolean">true</dhcp_hostname>
       **<hostname>localhost</hostname>**
       <resolv_conf_policy>auto</resolv_conf_policy>
       <write_hostname config:type="boolean">false</write_hostname>
     </dns>

Sorry for the trouble (and thank you again Darryl-G!) but hopefully this post can help someone else if they find themselves in the same issue.
